File: src/main/java/org/dxworks/githubminer/service/repository/GithubRepositoryService.java
package org.dxworks.githubminer.service.repository;

import com.google.common.collect.ImmutableMap;
import org.dxworks.githubminer.service.GithubApiService;
import org.dxworks.utils.java.rest.client.providers.BasicAuthenticationProvider;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

public class GithubRepositoryService extends GithubApiService {

    protected String owner;
    protected String repo;

    private ImmutableMap<String, String> map;

    public GithubRepositoryService(String owner, String repo) {
        super();
        this.owner = owner;
        this.repo = repo;

        initMap(owner, repo);
    }

    public GithubRepositoryService(String owner, String repo, BasicAuthenticationProvider authenticationProvider) {
        super(authenticationProvider);
        this.owner = owner;
        this.repo = repo;

        initMap(owner, repo);
    }

    private void initMap(String owner, String repo) {
        map = ImmutableMap.of("repo", repo, "owner", owner);
    }

    @Override
    public String getApiPath(Map<String, String> variablesValues, String... pathVariables) {
        Map<String, String> newMap = new HashMap<>(variablesValues);
        newMap.putAll(map);

        return super.getApiPath(newMap,
                Stream.concat(Stream.of("repos", ":owner", ":repo"), Arrays.stream(pathVariables)).toArray(String[]::new));
    }

    @Override
    public String getApiPath(String... pathVariables) {
        return super.getApiPath(map,
                Stream.concat(Stream.of("repos", ":owner", ":repo"), Arrays.stream(pathVariables)).toArray(String[]::new));
    }
}

File: github-miner-core/src/main/kotlin/org/dxworks/githubminer/service/repository/GithubRepositoryService.kt
package org.dxworks.githubminer.service.repository

import com.google.common.collect.ImmutableMap
import org.dxworks.githubminer.constants.ANONYMOUS
import org.dxworks.githubminer.constants.GITHUB_API_PATH
import org.dxworks.githubminer.http.factory.GithubHttpClientFactory
import org.dxworks.githubminer.service.GithubApiService

open class GithubRepositoryService(
    protected var owner: String,
    protected var repo: String,
    githubBasePath: String = GITHUB_API_PATH,
    githubTokens: List<String> = listOf(ANONYMOUS),
    clientFactory: GithubHttpClientFactory? = null
) : GithubApiService(githubBasePath, githubTokens, clientFactory) {

    private var map: ImmutableMap<String, String> = ImmutableMap.of("repo", repo, "owner", owner)

    override fun getApiPath(variableValues: Map<String, String>, vararg pathVariables: String): String {
        val newMap: MutableMap<String, String> = HashMap(variableValues)
        newMap.putAll(map)
        return super.getApiPath(newMap, *(arrayOf("repos", ":owner", ":repo") + pathVariables))
    }

    override fun getApiPath(vararg pathVariables: String): String {
        return super.getApiPath(map, *(arrayOf("repos", ":owner", ":repo") + pathVariables))
    }
}
